%-------------------------------------------------------------------------------
%	Packages and other Document configurations
%-------------------------------------------------------------------------------
\documentclass[a4paper,11pt]{article}
% Package declaration
%-------------------------------------------------------------------------------
% Specify input encoding
\usepackage[utf8]{inputenc}
% For A4 paper set all margins to 3cm
\usepackage[paper=a4paper,left=1.5cm,top=2cm,right=1.5cm,bottom=2cm]{geometry}%
% Set linespace, usage \doublespacing \singlespacing \onehalfspacing
\usepackage{setspace}%
% Set palatino font with small caps as default
\usepackage[sc]{mathpazo}%
% Rotation tools, including rotated full-page floats.
\usepackage{rotating}%
% Create subfigures
\usepackage{subfigure}%
% Extensive support for hypertext in LaTeX
\usepackage{hyperref}%
% For adding bookmarks to the document
\usepackage{bookmark}%
% For adding time to the document
\usepackage{datetime}
% For alignment of captions
\usepackage{caption}
% For multiple columns.
\usepackage{multicol}

% Start Article header
%-------------------------------------------------------------------------------
% Title
\title{Summary of QTL mapping population}%
% Authors
\author{\vspace{-5ex}}
%-------------------------------------------------------------------------------
% Dates
\date{\vspace{-5ex}}
%-------------------------------------------------------------------------------
% End article header

% For left aligning captions
\captionsetup{justification=raggedright,singlelinecheck=false}

% Start Document
%-------------------------------------------------------------------------------
\begin{document}

% Load required packages
<<label=setup, include=FALSE, cache=FALSE>>=
## load required packages
lapply(list("qtl", "xtable"), library, character.only = TRUE)
@

% Article title and date.
\maketitle
% Start single line spacing
\singlespacing

%-------------------------------------------------------------------------------
\section{Summary}
%-------------------------------------------------------------------------------
<<label=summary, echo=FALSE, results="asis">>=
summ <- summary(x)
summTab <- data.frame(c("Cross type", NA, "Number of individuals",
                        NA, "Number of phenotypes",
                        "Percent phenotyped", NA, "Number of chromosomes",
                        if (!is.null(summ$autosomes)) "  Autosomes",
                        if (!is.null(summ$Xchr)) "  X chr", NA,
                        "Total markers", "Number of markers",
                        "Percent genotyped", "Genotypes(%)",
                        if (!is.null(summ$typing.freq.A)) "Autosomes",
                        if (!is.null(summ$typing.freq.X)) "X Chromosome"),
                      c(switch(summ$type, "f2" = "F2 intercross",
                               "bc" = "Backcross",
                               "4way" = "4-way cross",
                               "riself" = "RI strains via selfing",
                               "risib" = "RI strains via sib matings",
                               "dh" = "Double haploids",
                               "haploid" = "Haploids"), NA,
                        summ$n.ind, NA, summ$n.phe,
                        round((1 - summ$missing.phe) * 100, 1), NA,
                        summ$n.chr,
                        if (!is.null(summ$autosomes)) paste(summ$autosomes, collapse = " "),
                        if (!is.null(summ$Xchr)) summ$Xchr, NA, sum(summ$n.mar),
                        paste(summ$n.mar, collapse = " "),
                        round((1 - summ$missing.gen) * 100, 1),
                        if (!is.null(summ$typing.freq.X)) {
                          roundedX <- sprintf("%.1f", summ$typing.freq.X * 100)
                          genoX <- paste(names(summ$typing.freq.X), roundedX, sep = ":")
                          if (!is.null(summ$typing.freq.A)) {
                            roundedA <- sprintf("%.1f", summ$typing.freq.A * 100)
                            genoA <- paste(names(summ$typing.freq.A), roundedA, sep = ":")
                            paste(genoA, collapse = " ")
                          }
                          paste(genoX, collapse = " ")
                        } else {
                          geno <- paste(names(summ$typing.freq),
                                        sprintf("%.1f", summ$typing.freq * 100), sep = ":")
                          paste(geno, collapse = " ")
                        }))

print(xtable::xtable(x = summTab, label = "summary",
                     align = c("l", "l", "l")),
      latex.environments = "flushleft", hline.after = NULL,
      include.rownames = FALSE, include.colnames = FALSE)
@
<<label=missPlots, echo=FALSE, out.width='90%'>>=
plot(x)
@
<<label=sumPlots, echo=FALSE, out.width='90%', out.height='.8\\linewidth'>>=
# Missing genotype information (markers / individuals)
op <- par(mfcol = c(1, 2))
hist(qtl::nmissing(x, what = "mar"), breaks = 20,
     xlab = "Number of missing values", main = "Missing values per marker")
hist(qtl::nmissing(x, what = "ind"), breaks = 20,
     xlab = "Number of missing values", main = "Missing values per individual")
par(op)
@

%-------------------------------------------------------------------------------
\section{Segregation distortion}
%-------------------------------------------------------------------------------
<<label=importance, echo=FALSE, results="asis">>=
## Evaluate the segregation distortion at every marker
## (and print those that show some evidence)
segr <- qtl::geno.table(x)
segr <- segr[segr$P.value < 1e-3, ]
if (nrow(segr) > 0) {
  cat("Markers that show evidence of segregation distortion.\n")
  print(xtable::xtable(x = segr, label = "segrDist",
                       align = c("l", rep("r", ncol(segr))),
                       digits = c(rep(0, ncol(segr)), -2)),
        latex.environments = "flushleft", hline.after = NULL,
        include.rownames = TRUE, include.colnames = TRUE)
} else {
  cat("No markers show evidence of segregation distortion.\n")
}
@
\clearpage
%-------------------------------------------------------------------------------
\section{Marker recombinations}
%-------------------------------------------------------------------------------
Recombination frequencies\\
<<label=recom, echo=FALSE, results="asis">>=
## Estimate recombination frequencies between all pairs of markers
crossRec <- qtl::est.rf(x)
qtl::checkAlleles(crossRec)
@
<<label=recomPlot, echo=FALSE, out.width='85%'>>=
## Plot of recombination frequencies
qtl::plot.rf(crossRec, alternate.chrid = TRUE)
@
\newpage
<<label=recom2, echo=FALSE, results="asis">>=
## Check number of recombinations per chromosome and per individual
cat("Number of recombinations per chromosome per individual\n")
NRc <- qtl::countXO(x, bychr = TRUE)
cat(paste("\n\nMean number of recombinations per chromosome per individual:",
          round(mean(NRc), 2)))
NRc <- data.frame(NRc, check.names = FALSE)
NRc$mean <- rowMeans(NRc)
NRc <- NRc[order(NRc$mean, decreasing = TRUE), ]
print(xtable::xtable(x = NRc[1:(ceiling(nrow(NRc) / 10)), ],
                     caption = "Individuals with highest number of
                     recombinations per chromosome",
                     label = "recInd",
                     align = c("l", rep("r", ncol(NRc))),
                     digits = c(rep(0, ncol(NRc)), 2)),
      latex.environments = "flushleft", hline.after = NULL,
      include.rownames = TRUE, include.colnames = TRUE,
      caption.placement = "top")
@
%-------------------------------------------------------------------------------
% End Document
\end{document}
