%-------------------------------------------------------------------------------
%	Packages and other Document configurations
%-------------------------------------------------------------------------------
\documentclass[a4paper,11pt]{article}
% Package declaration
%-------------------------------------------------------------------------------
% Specify input encoding
\usepackage[utf8]{inputenc}
% For A4 paper set all margins to 3cm
\usepackage[paper=a4paper,left=1.5cm,top=2cm,right=1.5cm,bottom=2cm]{geometry}%
% Set linespace, usage \doublespacing \singlespacing \onehalfspacing
\usepackage{setspace}%
% Set palatino font with small caps as default
\usepackage[sc]{mathpazo}%
% Rotation tools, including rotated full-page floats.
\usepackage{rotating}%
% Create subfigures
\usepackage{subfigure}%
% Extensive support for hypertext in LaTeX
\usepackage{hyperref}%
% For adding bookmarks to the document
\usepackage{bookmark}%
% For adding time to the document
\usepackage{datetime}
% For alignment of captions
\usepackage{caption}
% For multiple columns.
\usepackage{multicol}


% Graphics Path definition
%-------------------------------------------------------------------------------
\graphicspath{{figures/}}

%-------------------------------------------------------------------------------
%	Article Information
%-------------------------------------------------------------------------------
% Setting PDF properties in the final PDF
% !! Works only with PDFLaTeX and LaTeX->DVI->PS->PDF
\hypersetup
{
pdfsubject={Cross report},
pdftitle={Cross report}
}

% Start Article header
%-------------------------------------------------------------------------------
% Title
\title{Cross report}%
% Authors
\author{\vspace{-5ex}}
%-------------------------------------------------------------------------------
% Dates
\date{\vspace{-5ex}}
%-------------------------------------------------------------------------------
% End article header

% For left aligning captions
\captionsetup{justification=raggedright,singlelinecheck=false}

% Start Document
%-------------------------------------------------------------------------------
\begin{document}

% Set knitr options
<<label=setup, include=FALSE, cache=FALSE>>=
## load required packages
lapply(list("qtl", "xtable"), library, character.only = TRUE)
## Set global chunk options
knitr::opts_chunk$set(dev = c("pdf"),
                      fig.path = "figures/fig-",
                      fig.align = "center",
                      fig.show = "hold",
                      tidy = TRUE)
@

% Article title and date.
\maketitle
% Start single line spacing
\singlespacing

%-------------------------------------------------------------------------------
\section{Summary}
%-------------------------------------------------------------------------------

<<label=summary, echo=FALSE, results="asis">>=
summary(x)
@
<<label=sumPlots, echo=FALSE>>=
plot(x)
# Missing genotype information (markers / individuals)
op <- par(mfcol = c(1, 2))
hist(qtl::nmissing(x, what = "mar"), breaks = 20,
     xlab = "Number of missing values", main = "Missing values per marker")
hist(qtl::nmissing(x, what = "ind"), breaks = 20,
     xlab = "Number of missing values", main = "Missing values per individual")
par(op)
@

%-------------------------------------------------------------------------------
\section{Segregation distortion}
%-------------------------------------------------------------------------------

<<label=importance, echo=FALSE, results="asis">>=
## Evaluate the segregation distortion at every marker
## (and print those that show some evidence)
segr <- qtl::geno.table(x)
segr <- segr[segr$P.value < 1e-3, ]
if (nrow(segr) > 0) {
  cat("Markers that show evidence of segregation distortion.\n")
  print(xtable::xtable(x = segr, label = "segrDist",
                       align = c("l", rep("r", ncol(segr))),
                       digits = c(rep(0, ncol(segr)), -2)),
        latex.environments = "flushleft", hline.after = NULL,
        include.rownames = TRUE, include.colnames = TRUE)
} else {
  cat("No markers show evidence of segregation distortion.\n")
}
@
\newpage
%-------------------------------------------------------------------------------
\section{Marker recombinations}
%-------------------------------------------------------------------------------
Recombination frequencies\\
<<label=recom, echo=FALSE, results="asis">>=
## Checking marker order by first estimating recombination frequencies
## between all pairs of markers
crossRec <- qtl::est.rf(x)
qtl::checkAlleles(crossRec)
@
<<label=recomPlot, echo=FALSE>>=
## Plot of recombination frequencies
qtl::plot.rf(crossRec, alternate.chrid = TRUE)
@
\newpage
Number of recombinations per chromosome per individual\\
<<label=recom2, echo=FALSE, results="asis">>=
## Check number of recombinations per chromosome and per individual
NRc <- qtl::countXO(x, bychr = TRUE)
cat(paste("\n\nMean number of recombinations per chromosome per individual:",
          round(mean(NRc), 2)))
NRc <- data.frame(NRc, check.names = FALSE)
NRc$mean <- rowMeans(NRc)
NRc <- NRc[order(NRc$mean, decreasing = TRUE), ]
print(xtable::xtable(x = NRc[1:(ceiling(nrow(NRc) / 10)), ],
                     caption = "Individuals with highest number of
                     recombinations per chromosome",
                     label = "recInd",
                     align = c("l", rep("r", ncol(NRc))),
                     digits = c(rep(0, ncol(NRc)), 2)),
      latex.environments = "flushleft", hline.after = NULL,
      include.rownames = TRUE, include.colnames = TRUE,
      caption.placement = "top")
@

\clearpage
%-------------------------------------------------------------------------------
\section{Crossover points}
<<label=crossover, echo=FALSE>>=
## Display marker genotypes to show crossover points
crossErrLod <- qtl::calc.errorlod(x)
qtl::plot.geno(crossErrLod, chr = 1, c(1:5))
@
%-------------------------------------------------------------------------------
%-------------------------------------------------------------------------------
% End Document
\end{document}
