<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Bart-Jan van Rossum" />

<meta name="date" content="2018-04-18" />

<title>Use of RAP package</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Use of RAP package</h1>
<h4 class="author"><em>Bart-Jan van Rossum</em></h4>
<h4 class="date"><em>2018-04-18</em></h4>



<div id="general-introduction-about-package" class="section level1">
<h1><span class="header-section-number">1</span> General introduction about package</h1>
<p>TO BE WRITTEN</p>
<hr />
</div>
<div id="create-td-object" class="section level1">
<h1><span class="header-section-number">2</span> Create TD object</h1>
<p>The first step when modeling a field trial or performing a GxE analysis using the RAP package will always be creating an object of class TD. Such an object is used internally by the package for performing all types of analysis.</p>
<p>A TD object can be easily created from a data.frame by calling the function <code>createTD</code>. This function will do a several things:<br />
* Rename columns to default column names used by the RAP package. For example, the column in the data containing variety/accession/genotype will allways be renamed to genotype. Original column names will be stored as an attribute to the final object of class TD.<br />
* Convert column type to the default column types. For example the column genotype will always be converted to a factor and rowCoord will always be converted to a numeric column.<br />
* Split the data by trial. The output TD object will be a list of data.frames where each data.frame contains the data for a single trial. If there is only one trial or no column trial is defined the output will be a list with one item only.<br />
* Add meta data to the seperate data.frames in the TD object. This meta data consists of location, date, longitude, latitude, plot design, width and length. None of this is strictly neccessary for doing analysis but it is used for plotting field layouts, plotting trials on a map and naming plots. For a single trial can simply be added when creating the TD object using the function parameters in <code>createTD</code>. However if the data consists of multiple trials with different meta data per trial it is more convenient to first create a TD object without using the parameters for adding meta data. Doing so only a location will be added based on the name of the trial. After creating the TD object the meta data can be extracted using <code>getMeta</code>. This function outputs a data.frame with the meta data per trial. The content of this data.frame can then be modified and added back to the TD object using <code>setMeta</code>.</p>
<p>After creating a TD object data for new trials can be added to it using <code>addTD</code>. This function works in exactly the same way as <code>createTD</code> except that it adds data to an existing TD object instead of creating a new one. Still the output needs to be assigned to a variable.</p>
<p>As an example we will create a TD object using field data from an experiment with wheat in Chili described by <span class="citation">Lado et al. (2013)</span>. This data will be used as an example throughout this vignette.</p>
<p>The experiment was performed in two locations with 2 different drought regimes in two years for the first location and 1 trial for the second location. For the sake of the example we will create an TD object for the first location first and add the data for the second location later on. In practice this could also be done in one go.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;wheatChl&quot;</span>)
wheatTD &lt;-<span class="st"> </span><span class="kw">createTD</span>(<span class="dt">data =</span> wheatChl[wheatChl<span class="op">$</span>trial <span class="op">!=</span><span class="st"> &quot;C_SWS_12&quot;</span>, ], 
                    <span class="dt">genotype =</span> <span class="st">&quot;trt_id&quot;</span>, <span class="dt">repId =</span> <span class="st">&quot;rep&quot;</span>, <span class="dt">subBlock =</span> <span class="st">&quot;bl&quot;</span>, 
                    <span class="dt">rowId =</span> <span class="st">&quot;row&quot;</span>, <span class="dt">colId =</span> <span class="st">&quot;col&quot;</span>, <span class="dt">rowCoord =</span> <span class="st">&quot;row&quot;</span>, 
                    <span class="dt">colCoord =</span> <span class="st">&quot;col&quot;</span>)</code></pre></div>
<p>Note that both row and col are used twice. This is done since whenever a spatial model is fitted both a factor column containing the spatial information and a numeric column containing this information are needed. <code>rowId</code> and <code>colId</code> will be converted to a factor and <code>rowCoord</code> and <code>colCoord</code> will remain numerical column as they were in the input.</p>
<p>The TD object we just created will be a list with 4 items, 1 for each trial (combination of location, drought regime and year) in the original data.frame. The meta data will be a data.frame with 4 rows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wheatMeta &lt;-<span class="st"> </span><span class="kw">getMeta</span>(<span class="dt">TD =</span> wheatTD)
<span class="kw">head</span>(wheatMeta)
<span class="co">#&gt;           trLocation trDate trDesign trLat trLong trPlotWidth trPlotLength</span>
<span class="co">#&gt; SR_FI_11    SR_FI_11   &lt;NA&gt;       NA    NA     NA          NA           NA</span>
<span class="co">#&gt; SR_FI_12    SR_FI_12   &lt;NA&gt;       NA    NA     NA          NA           NA</span>
<span class="co">#&gt; SR_MWS_11  SR_MWS_11   &lt;NA&gt;       NA    NA     NA          NA           NA</span>
<span class="co">#&gt; SR_MWS_12  SR_MWS_12   &lt;NA&gt;       NA    NA     NA          NA           NA</span></code></pre></div>
<p>Now we can modify the meta data and then add it back to the original TD object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wheatMeta<span class="op">$</span>trLocation &lt;-<span class="st"> &quot;Santa Rosa&quot;</span>
wheatMeta<span class="op">$</span>trDate &lt;-<span class="st"> </span><span class="kw">as.Date</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;310811&quot;</span>, <span class="st">&quot;310812&quot;</span>), <span class="dt">times =</span> <span class="dv">2</span>), <span class="st">&quot;%d%m%y&quot;</span>)
wheatMeta<span class="op">$</span>trLat &lt;-<span class="st"> </span><span class="op">-</span><span class="fl">36.32</span>
wheatMeta<span class="op">$</span>trLong &lt;-<span class="st"> </span><span class="op">-</span><span class="fl">71.55</span>
wheatMeta<span class="op">$</span>trPlotWidth =<span class="st"> </span><span class="dv">2</span>
wheatMeta<span class="op">$</span>trPlotLength =<span class="st"> </span><span class="dv">1</span>
wheatTD &lt;-<span class="st"> </span><span class="kw">setMeta</span>(<span class="dt">TD =</span> wheatTD, <span class="dt">meta =</span> wheatMeta)</code></pre></div>
<p>To add the final trial to the TD object the function <code>addTD</code> can be used. Since in this case we are only adding 1 new trial it makes sense to immediately add the meta data for this trial as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wheatTD &lt;-<span class="st"> </span><span class="kw">addTD</span>(<span class="dt">TD =</span> wheatTD, <span class="dt">data =</span> wheatChl[wheatChl<span class="op">$</span>trial <span class="op">==</span><span class="st"> &quot;C_SWS_12&quot;</span>, ], 
                 <span class="dt">genotype =</span> <span class="st">&quot;trt_id&quot;</span>, <span class="dt">repId =</span> <span class="st">&quot;rep&quot;</span>, <span class="dt">subBlock =</span> <span class="st">&quot;bl&quot;</span>, 
                 <span class="dt">rowId =</span> <span class="st">&quot;row&quot;</span>, <span class="dt">colId =</span> <span class="st">&quot;col&quot;</span>, <span class="dt">rowCoord =</span> <span class="st">&quot;row&quot;</span>, 
                 <span class="dt">colCoord =</span> <span class="st">&quot;col&quot;</span>, <span class="dt">trLocation =</span> <span class="st">&quot;Cauquenes&quot;</span>, 
                 <span class="dt">trDate =</span> <span class="kw">as.Date</span>(<span class="st">&quot;070712&quot;</span>, <span class="st">&quot;%d%m%y&quot;</span>), <span class="dt">trLat =</span> <span class="op">-</span><span class="fl">35.58</span>,
                 <span class="dt">trLong =</span> <span class="op">-</span><span class="fl">72.17</span>, <span class="dt">trPlotWidth =</span> <span class="dv">2</span>, <span class="dt">trPlotLength =</span> <span class="dv">1</span>)
## Inspect the meta data after the extra trial was added.
<span class="kw">getMeta</span>(<span class="dt">TD =</span> wheatTD)
<span class="co">#&gt;           trLocation     trDate trDesign  trLat trLong trPlotWidth</span>
<span class="co">#&gt; SR_FI_11  Santa Rosa 2011-08-31       NA -36.32 -71.55           2</span>
<span class="co">#&gt; SR_FI_12  Santa Rosa 2012-08-31       NA -36.32 -71.55           2</span>
<span class="co">#&gt; SR_MWS_11 Santa Rosa 2011-08-31       NA -36.32 -71.55           2</span>
<span class="co">#&gt; SR_MWS_12 Santa Rosa 2012-08-31       NA -36.32 -71.55           2</span>
<span class="co">#&gt; C_SWS_12   Cauquenes 2012-07-07       NA -35.58 -72.17           2</span>
<span class="co">#&gt;           trPlotLength</span>
<span class="co">#&gt; SR_FI_11             1</span>
<span class="co">#&gt; SR_FI_12             1</span>
<span class="co">#&gt; SR_MWS_11            1</span>
<span class="co">#&gt; SR_MWS_12            1</span>
<span class="co">#&gt; C_SWS_12             1</span></code></pre></div>
<p>To get an idea of the content of the data and the field layout of the data we can summarize and plot the TD object. Multiple traits can be summarized at once but in order to save space we only look at GY (grain yield).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(wheatTD, <span class="dt">trial =</span> <span class="st">&quot;SR_FI_11&quot;</span>, <span class="dt">traits =</span> <span class="st">&quot;GY&quot;</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Summary statistics for GY in SR_FI_11 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Number of observations  794</span>
<span class="co">#&gt;   Number of missing values  6</span>
<span class="co">#&gt;                       Mean  8015.6</span>
<span class="co">#&gt;                     Median  8021</span>
<span class="co">#&gt;                        Min  2239.5</span>
<span class="co">#&gt;                        Max  12992</span>
<span class="co">#&gt;             Lower quartile  6728.5</span>
<span class="co">#&gt;             Upper quartile  9321.75</span>
<span class="co">#&gt;                   Variance  3080361.436</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(wheatTD, <span class="dt">trials =</span> <span class="st">&quot;SR_FI_11&quot;</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAA2FBMVEUAAAAAADoAAGYAOjoAOpAAZrY6AAA6ADo6AGY6OgA6OpA6ZmY6ZrY6kNtNn/9Z/5xmAABmADpmAGZmOpBmZrZmtttmtv9wTf+AgICAuv+QOgCQOjqQOmaQZmaQZpCQkGaQ2/+ZgP+f/0Cm/8mysrKy1v+2ZgC2Zjq2Zma2/9u2//+//4DCsv/JzD3Q0ZLZ/7LbTf/bkDrbtrbb25Db29vb///mgP/m5ubwsv/7/039/7L/jIz/ok3/srL/tmb/trb/vYD/17L/2dn/25D//7b//8z//9v///871w1lAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAMiUlEQVR4nO2dC7vbtg2GlTQ+bbec7aTttO1ka8+6zWmz7qI12bqbdrFr//9/NIEUZJkCSdOgSJnC9zyNHUtWpLckCJMAUR3P9f0fSX1/pHR4/bT/jDxSjirj70GAjm8EkBvQ7u7Z0+y3mFcTQBZlubklaALot6QEEEoAGeIB2lbVZvZbzCtuC2rqmW8wt5iAih/luYCKb0BcQF+9nfsGc4sH6PD5+9nvMLPEUfRoAugbUgIIJYAMCSCPeIB2d9X97LeYVzxATV38OMbsYtv1zQcFAdrWh18U7iryAL15OrwuvAmJkfZIPGmPJoC+IyWAUALIEA/Q7u554YMYE9Cbp93L2W8xrwSQRzxAbfWBAHIb6XZ1flAQoKYSI+1pQcVLPGmPJoB+TUoAoQSQIRag/WfH/UNV9uIqB9DubnNs6v2nRU+6clvQV28LX34WQB5xAUkXEyN9LnEUDU0A/YOUAEIJIEMCyCMWIAjhLD2MkwMIPGn4r2hJC/JIAHkkgDwSQB6JJ+3RBNBvSAkglAAyxALU2Z/is1Y5gMBJfPPUlu0pclvQ8diubD4oFNDqhvlAQPtPip6R5gMqPq1XHEWPJoD+Q0oAoQSQIQHkEQtQZ6Tbam2edAAg8KTfuR3Fpur0wr30SixdHx4rrfwRftwWdGwdzzA8p4Ph/oG4wPDF/O2TDejoiJRuKn9D2FIHT4Cye1lcQM5NuraqAcDTQhPa3WGLaLqXrX6/7fmdDoK6r6h+uX/Qr+qoDjgeddrz78wkLiDndEffOnRWWXvqM9iyuvc9oNFBEALqDsNrf/7oLVzX+M5MmtWTbs6feYOsGvWEjWoUCuL4YH8ytpzuVf+pTu+a1EahuZ98ZyZNAP2O1HWA0JQMNrrRRqVRH3XPthnboOZkcUY26H44BV6675yZrNF3ZtKsgIZuAA+FD73RNujYt4ateRB0ArQ5NSf4Vn/gfnzSggGBffamPbf6IdTj1BrKBND4IAifHS31pr9QDX/pG6X5nZnEAaTW5R1+0GAf+s5xjw87ATQ+CNKNptGExi3oeBzGMfM7M4nbgg6f/9IKCP5v1/pVjVP1yUgbgMYHQT2TRvefkQ066jT0rXHBGcUF1NaOINfBUaw2eiiCbmG0IPV+fBCEjWaLEHAU02/VYfM7M4kLqHHd4Pgn1fDeAAQm6vnXo4P9F3GYh8bTnA42g/NzML4zk9hG2h0m3Z4POs+eGu0BnQAdlUs9Otifi46iOkf5zNpZUG/VacZ3ZtLMw/ztS+akPZoA+jkpAYQSQIZYgNSv+bWlIgQAAk+68JV5fgtC/79YcQEVH73ABVR8XQS+kV5bFxNH8VwTQL8iJYBQAsiQAPKIBWiVnnQAIB0GbH6K0+q4muH59xeniIC6FnR4/TE1zLdqjqwN8AAqOjYyvaIDohzFw6OeBm08gS8rAER70vsHbZdcoTHmfdBpWOkVGxDpSR8e+/WwgBaUGwzKCyiKJ92qoa0JsUH0lhjp5QX0F1KhfpAayi5vP9190JvLpVciQMEqE5AtFQFtUAggeqPv9IoJSG8qQBSIwlFs7YDs6VAB4zveB110Kb2iA6IqjA2/NghOW33E7IMVXb40vaIDCtzBDF2jrWG5igUUWJll/6p3jXYfnYOt/r4QpXEUrSHMBQD6A6lgTxpsTENMFWHXMn+FVP9diJIAQj+I+jXfWIx0bjCoJIACfs0P81T/WoiiAtJTrlOv2dGCIPvi27up+139byGKCUhvdkuNYw4bdN99DGGK5jCfGwwqdguyVBhzjWJqJJuMYn9diGIDCqwwtj5AoRXGcJifdLHfL0RJjLQj162335PJ2OrLhSjVlKvNSNt0O4B+TCqeo2gD9LOFKAmga5Z9coNBRQWkbBCZtBzegv60EMUEpNfmqbTwK2xQbjCo2C3IUmHMnrFtm1H86UIUG1BohTHrjOKPFqLYgAIrjNknzHKDQcUGRKVr6eCO3R1Bzg7obwtRgmF+/6C7Eb6eyTqjuBj5APE96eHZ9w9EG7LNKC5GPkD/JBUAaLTufIEflBvHVLMDGq07E540JJoCwIkNort2esUEBMFT09I1o47VTAGBDYI+OAFEzyKkV0xAsNEtUfzo5OJMI8z0KNa1rFUAOuptBczF591dbbwxAR3bF9+agOgV3fSKCshSX0z/EuvGK2IQ65vXxH2q6NjI9IoJCDa6JeuL6U0RyB0RhgjhNQCC6LJY9cUqOk8tvRIFLxQMiP6FEjqjGLzjSEVviZFeSQBhiFlInHRuMKg0gBBTQAgevT1hesUEpNIQbEnP0IiIFmSdMJsnNSVcMQGBJ03XF4MZV3JFwz4flBsMKm4Xa2uyBZF9ywModxILKioge9kIvYkkIeuEWW4wqJiAVMkI68YLDb1Lsm3CLHeWDyomIJWn4diZwt7TRv8+KncSCyqNJ92P5QH7r90OIHoxJNiTDt3R43aSWWIAsuuUM27+ms8NBhUTEDiKlt1fHEvP9Oh2O9k+AYDAUaTrizmCFyzJdkUC0rXFiF3Mrgl/yZ3lg4oKyOYoXhNAlTuJBRUTEDiKZH2xa1pQbjComIDAUaTri10RQJU7iQWVOV+sAEB03NX8c9K5s3xQSQBdkzefOYdlUBJA1+TN5waDigkIvGi6vtgVefO5k1hQMQGBF03WF3Plza8JkPKiwxsLDWgxignIV18snszbTHcGBxB40Yk2u71NQOBFJ9rsdkGAMgUveLQgQHSKhwBCCSDPZyGAlP0JSnq+WrcJSE23RvKDFiteF2trV32xIsQC5KsvVoI4gFSUa/5auvOKA0hPt0oLctigFehGPOl8mgCi0+0FEEoAGeIA0sEL81dFgDUk5xKkCng8PDomMNUZjuvAoZq8BgcQeNIp6otR4RFj6Ui27cYWOYJn2K8DqwxQooC4Bq+LtXWK+mJb9yJSU/2wax8QP2uGzp6f4bhOq6rb1tQ1WIDS1BfzbXH15/fwULsekv0Mz3W6VkRdgwMo0ZTr/tXHnmbqA3Tsjzqv07x4HxtQoinX3YdPdBDJ6YyLALmvAxl+FwFapqNof3TQZYCc11EVHC4CRO/stXxALiPtB6QzRC8y0ssDBFNyux94/SDXMK/OcFwHeh/okmF+eYBgSsU9bakA7R98jqL9OjrwvaauwQGUypPOKg4gHQac4CZzitfF2p/Q9cUKEguQrb5YSeIAShi8kE8cQAmDF/LpRjzpfJoAondZFkAoAWSIAwjsD52KUJA4gHSuGFUeqiDxuphlY4GSxAJkqy9WkjiA1PRc4UvzLEBgoAPri92exFH0aAKIzmBYJiCdVaT23PMmf4aULh/r9gHp9E+6xNBIqwXUb/d5ePS0oRyA+inX4OzC6wV764Dj3lZqgx0AhJnV/357+vx0wqsvYJK5+9qzLzIA0pvdJhzHYPEKNhtu1fz6BgDhzp5K+PnoBMC37RjRiX8XiNfFOk/a2/njCTdi1imxLezVXI93PR8+H51wj3vuXnufLEDh9cVYwt2F9Wv3pwFo+Hx8AuaH5rBBeso1rL4YR7i8p59f0TnrYsPn4xNw1TQHID3lms5IUy1oMNKd2UnTghbsSVM2aDzMUzaoRp5NLEB0NvkiAKmHhD41GsV6R/HwCM9PjGLIJt4otmRAlB/UJ6Hrjjbxg/QOB1H9oABA2lFcWzpUACC9Np8kLTyfeF2srQWQC9AV9cVuThxAOiIyqL7Y7YkDCBxFqr5YUbqtYT6DbsmTzqIJIHpnLwGEEkCGOIDg902s0jWLFQcQbM9FFj8qSawu1j5/S5XPKko8G7R7KYAcgCDCVbqYAxBdBLIwiaPo0QQQvUeuAEIJIEMCyCMOILVSsLogzgBA4EnT9cUKEquLwZqltCCXDdq9FEAOQCpXTADZATmLQJYi8aQ9mgCi65UIIJQAMsQBBI7i+nJWAwC9U2vzvg20blysLqaC21LsYpZRPBu0RkcxABA4iu6toQoQB9A6k3rFUTzXBBBdGlEAoQSQIQHkEQcQeNJXB2jfijiAwJMm64uVJFYXA096dXU1gmwQxEivLU46ABB40uvboisAkKOcejkST9qjCSC6CrsAQgkgQxxAyklMl/ScRxxA7yLWWV2sWF2sg7O++mJBNmj3cn31xQIAqSjX1c0oBgDSTqK0IEcXW4HEk/ZoAugbUgIIJYAMcQDp4IW1pSIEAAJPuvAJV2YX6zzpFPXFsopng2C6dW0zigGA1jnlGgBIplzFUSQAfUdKAKEEkCEB5BEH0Do96QBAOgw4wU3mFKuLtc+/Lr6+2P8BNAJCP84TVy4AAAAASUVORK5CYII=" /><!-- --></p>
<p>The default plot creates plots for the layout of all trials in the TD object. This can be restricted to selected trials using the <code>trials</code> parameter. The width and length of the plot will be derived from the trPlotWidth and trPlotLength in the meta data if these are available.</p>
<p>A second type of plot will plot the trial locations on a map. This plot is made based on trLat and trLong in the meta data. If not available the corresponding location will not be plotted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(wheatTD, <span class="dt">plotType =</span> <span class="st">&quot;map&quot;</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABdFBMVEUAAAAAADoAAGYAALkAANsAAP8AJ5AAKJ4AOpAAOv8APpYAPrkAUskAZrYAZv8AeNsAhJYAkMUAkP8AttAAtv8AvsAAxP8A0dAA27YA2/8A39AA4cwA/+YA/+4A//MA//8D2/8D39AD4dwD/+gD/+8EZv8EeZ4E//8GcrEIkOEIvv8I/+gI//MKALkLALkLANsLJ5ALKJ4LUrELZv8NZtsTOtsZw/8Z3/kcAP8dAP8dOgAhANshAP8hb/8hif8ktP8r//8tJwAtvv86AAA6kNs66Nw68NA6//88Y/884f88//8/KNs/KP8/pP8/tP9D//9L//9d//9mAABmtv9m4f9m//9qPv9q//92yf+PUtuP//+QOgCQOjqQtpCQ29uQ2/+Q/7GQ/+6Q//+mUtum/5am//+2ZgC2Zma24f+2/9u2///bkDrb/9vb////b7n/pJ7/pMD/tND/tmb/w+D/25D//5b//7H//7b//8z//9v//93////Nxke0AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAG8klEQVR4nO3diXsTRQCG8SmwgEjBo3igsN73Bd4CXlSrKIjWUMWjGkzVxrZSLALzzzvZ2aTbJpsvaGbM9HnfhydNk92d5MfuZJNWNJaGZv7vBzDpASQCSASQCCARQCKARACJABIBJAJIBJAIIBFAIoBEAIkAEgEkAkgEkAggEUAigEQAiQASASQCSASQCCARQCKARACJxgy0ZHpNd2+7+Z3Zv1JZZvP77ffU9+vybSw81pIA+vurXTsZaHu3D7RkHND/VIA5yD1tj9M2Uz+cNVM/eoa/zjq1qcVBQMVds37t9Q7uvLWVFQr1Xb/3Ley2v7jkF+l84wqwi4UFKh70H8UzW/f7lXs2fUB+OX9juQvO2soKW4AqC5dXC6Hy+vj3tMBAnedRMPgb3d/+bB+Qu9Fdtrsqs27KcU+0soI/xPoXbhcg7XKV6XLtMRcYaN5WD6l2MTVtBypnmOKLv97u7gh+hSpQdWG/fUc23bkINE0FnoMWbZfBXZZz9zag7redpauz9uYKFaDqwuX2i52nXHjs+088oOIZzHefTBXIHx3F8THfW7X08StUgKoLV4E6l6Y3vY+zWEB+MhkEVLMHVVYYYQ/yw4V4HYsFtN6dL/on6f45aN2tV1mhfg7aBLr5/aINcsIUcQ/yx1n/JL3tVcxPKPOVFfyGBr2K9YD8zSFOtqPOQQMn6d45TLHSUmWp3s3r/edB03YLUPdmf9I4zqK+ik0ttv0ZzqAz6XJ6Xe++Gm2uULB1T8grC2+Zg4qbA7zU83GHCCARQCKARACJABIBJAJIBJAIIBFAIoBEAIkAEgEkAkgEkAggEUAigEQAiQASASQCSASQCCARQCKARACJABIBJAJIBJAofaDAzyB5IAPQ0EzopwCQHiDpABIZG/g5ADTCACkHkAggEUAigESmdxFygIQDSASQCCCRqVwGHCDdABKVQOE+FdoZQAGfB0AjDpBqZsuXcAOkWvAdCKBRR0i04EfYzgDiTLo2gEQADS/8FLQjgPhEsTaAROkCRXIHaKRhUgSKFEDDi+AD0GhjJFrwH2mE33zYABIBJAr+mwsRth80gEQAiYL/iqtNAqiZ543B9wDUqTlnbaM58C6AOs2tWLuyYlfyPC++aTbtcp43G+76VLNprL+90bkM0eQD3Z9l2ZmCprHsgW4dvexEPNDBR7Js5i47t9y5N0QTD3Tt3gW79vJ5J+R2Gw/021Mtu1wCHXY7UP7sBb9nhWjiga4+ern4ejzLjjWvPvZWduynDw/ta31z6I6HVnY1moePtKy99O7ct9mDzwQZf+KB7Kf37F5ozF070rp1/NzVxy/eeOHctaOX/3z6ixvPX9w/1zz4xJe2+dqFj19c+Pn1IMNPPpB18+/dC+5yJn9j7bn84Tffc5f5109mM/kDjebeV/N85qT96IBtfhJk9ASAXFdOru5e2Dh1fu1E6/pnncvVmfc3Tn3g7tl7ouVut6sH3IEWZOiJB3Iy9vrnZ5zA6p4K0H2N1T0doLVXWnbtpQX3Z+NUkPEnHshecS/zJ+3G29mdp894oI3T+3457b59x3oge2lf60pxLhCgyQeqLcYHrgCNOEqSASSK8YFrykD+3SlAdeXufXyMh58wUAbQsIw7xDKA6jOZtVmER58ukM1i7EDJAsU5CYozRJAAEsU5S4w0RogAGl60IyxtIOag2gAScYiJABIBJAJIBJAIoOHFe5VPGijKYwdopKFSy3QCqD5T/PomQLUBJAJIVExAcR56okDRdqAkgQxAwyuBmIOGFeO/hOoNlF7+300M/b/U8ENFGGPMmZgn0ukCWQ6x+gAanun+063MQYMDaHim70qU0ZLJDLgWZbhUMuxBqogzUJpAMT+SThHIlLsQQIMzxr8Vi/TI0wQqvkYaLs4wYyzm+4yI44wv0/28LNJwsQYaXxE/67ApAnUn6VjDRRtpXMXdgVIEsgANjUlaVH4kHWsaSg8o7luxNIE4kx6a/82FWMdYgkC9OSgKUpJAdnOSDo4U4a+gpv+4zcqPn4MSxdtHzcBv/+U2y0MswmSUItAmjAn/Wx4pAnXfjwE0bKOmtycBNHCrvWmIPah+y70DLWApA5U/3eBVrH7T1VPGQCUN1BMK+CFjim81thV2F9oBQDao0c4AChhAIoBEAIkAEgEkAkgEkAggEUAigEQAiQASASQCSASQCCARQCKARACJABIBJAJIBJAIIBFAIoBEAIkAEgEkAkgEkAggEUAigEQAiQASASQCSASQCCARQCKARACJABIBJAJIBJAIIBFAIoBEAIkAEgEkAkgEkAggEUAigEQAiQASASQCSASQCCARQCKARACJABIBJAJIBJAIIBFAIoBEAIkAEgEkAkgEkOgfittNmljOBTkAAAAASUVORK5CYII=" /><!-- --></p>
<hr />
</div>
<div id="modeling" class="section level1">
<h1><span class="header-section-number">3</span> Modeling</h1>
<p>After creating a TD object the next step will usually be fitting a model on the trial data. This can be done using the function <code>STRunModel</code>. Depending on the design of the trial several different models can be fitted. The design can be given as a parameter in the function or can be included in the meta data of the TD object. In the first case the same model will be fitted for all trials, in the latter case different models can be fitted for different trials depending on the design.</p>
<p>All models can be fitted using 3 different model fitting engines, SpATS <span class="citation">(Rodríguez-Álvarez et al. 2017)</span>, lme4 <span class="citation">(Bates et al. 2015)</span> and asreml. For models with rowcol or res.rowcol design SpATS is the default engine, for the other models lme4. This can be overruled by the function parameter <code>engine</code>.</p>
<p>The exact models fitted are the following<br />
* ibd trait = genotype + <em>subBlock</em> + e<br />
* res.ibd trait = genotype + <strong>repId</strong> + <strong>repId:subBlock</strong> + e<br />
* rcbd trait = genotype + <strong>repId</strong> + e<br />
* rowcol trait = genotype + <em>rowId</em> + <em>colId</em> + e<br />
* res.rowcol trait = genotype + <strong>repId</strong> + <em>repId:rowId</em> + <em>repId:colId</em> + e</p>
<p>In the models above fixed effects are indicated in <em>italics</em> whereas random effects are indicated in <strong>bold</strong>. genotype will be fitted as fixed or as random effect, or both, depending on the value of the parameter <code>what</code>. Extra fixed effects may be fitted using the parameter <code>covariates</code>.<br />
In case SpATS is used as engine for fitting models an extra spatial term is always included in the model. The same is done when the engine is asreml and the function parameter <code>trySpatial</code> is set to true.</p>
<p>Using the TD object wheatTD we created in the previous section we can now fit a model on the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">modWheat &lt;-<span class="st"> </span><span class="kw">STRunModel</span>(<span class="dt">TD =</span> wheatTD, <span class="dt">trials =</span> <span class="st">&quot;SR_FI_11&quot;</span>, <span class="dt">traits =</span> <span class="st">&quot;GY&quot;</span>,
                       <span class="dt">design =</span> <span class="st">&quot;res.rowcol&quot;</span>)</code></pre></div>
<p>Note that by not supplying the <code>what</code> argument to the function two models will be fitted, one with genotype as a fixed effect and one with genotype as a random effect. Because of this both BLUEs and BLUPs can be calculated from the output and this will be done when summarizing the results. If genotype would have been fitted as fixed effect only, the summary would only have contained BLUEs. When fitting genotype as random effect only the summary would only have contained the BLUPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Set nBest to 5 to decrease size of output.
<span class="kw">summary</span>(modWheat, <span class="dt">nBest =</span> <span class="dv">5</span>)
<span class="co">#&gt; Summary statistics: </span>
<span class="co">#&gt; ===================</span>
<span class="co">#&gt;                                 GY</span>
<span class="co">#&gt; Number of observations       794.0</span>
<span class="co">#&gt; Number of missing values       6.0</span>
<span class="co">#&gt; Mean                        8015.6</span>
<span class="co">#&gt; Median                      8021.0</span>
<span class="co">#&gt; Min                         2239.5</span>
<span class="co">#&gt; Max                        12992.0</span>
<span class="co">#&gt; Lower quartile              6728.5</span>
<span class="co">#&gt; Upper quartile              9321.8</span>
<span class="co">#&gt; Variance                 3080361.4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Estimated heritability </span>
<span class="co">#&gt; ======================</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Heritability: 0.65 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Predicted means (BLUEs &amp; BLUPs) </span>
<span class="co">#&gt; ===============================</span>
<span class="co">#&gt; Best 5 genotypes</span>
<span class="co">#&gt;         BLUEs       SE    BLUPs     SE</span>
<span class="co">#&gt; G369 10649.75   681.05  9643.18 542.30</span>
<span class="co">#&gt; G349 10351.81   668.15  9563.50 537.49</span>
<span class="co">#&gt; G341 10294.38   665.01  9468.74 536.49</span>
<span class="co">#&gt; G227 10043.79   673.05  9351.96 540.84</span>
<span class="co">#&gt; G271 10010.28   663.48  9398.66 536.11</span></code></pre></div>
<hr />
</div>
<div id="extract" class="section level1">
<h1><span class="header-section-number">4</span> Extract</h1>
<p>Explanation on what can be extracted from a model and how</p>
<hr />
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-Bates2015">
<p>Bates, Douglas, Martin Mächler, Ben Bolker, and Steve Walker. 2015. “Fitting Linear Mixed-Effects Models Using Lme4.” <em>Journal of Statistical Software</em> 67 (1). doi:<a href="https://doi.org/10.18637/jss.v067.i01">10.18637/jss.v067.i01</a>.</p>
</div>
<div id="ref-Lado2013">
<p>Lado, Bettina, Ivan Matus, Alejandra Rodríguez, Luis Inostroza, Jesse Poland, François Belzile, Alejandro del Pozo, Martín Quincke, Marina Castro, and Jarislav von Zitzewitz. 2013. “Increased Genomic Prediction Accuracy in Wheat Breeding Through Spatial Adjustment of Field Trial Data.” <em>G3: Genes<span class="math inline">\(\vert\)</span>Genomes<span class="math inline">\(\vert\)</span>Genetics</em> 3 (12): 2105–14. doi:<a href="https://doi.org/10.1534/g3.113.007807">10.1534/g3.113.007807</a>.</p>
</div>
<div id="ref-RodAlv2017">
<p>Rodríguez-Álvarez, María, Martin P. Boer, Fred Eeuwijk, and Paul H.C. Eilers. 2017. “Correcting for Spatial Heterogeneity in Plant Breeding Experiments with P-Splines.” <em>Spatial Statistics</em> 23 (October): 52–71. doi:<a href="https://doi.org/10.1016/j.spasta.2017.10.003">10.1016/j.spasta.2017.10.003</a>.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
